neo.UnitTests/Cryptography/ECC/UT_ECDsa.cs

```diff
@@ -0,0 +1,85 @@
+﻿using FluentAssertions;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using Neo.Cryptography.ECC;
+using Neo.Wallets;
+using System;
+using System.Numerics;
+using System.Reflection;
+using ECDsa = Neo.Cryptography.ECC.ECDsa;
+using ECPoint = Neo.Cryptography.ECC.ECPoint;
+
+namespace Neo.UnitTests.Cryptography
+{
+    [TestClass]
+    public class UT_ECDsa
+    {
+        private KeyPair key = null;
+
+        [TestInitialize]
+        public void TestSetup()
+        {
+            key = UT_Crypto.generateCertainKey(32);
+        }
+
+        [TestMethod]
+        public void TestCalculateE()
+        {
+            ECDsa sa = new ECDsa(key.PublicKey);
+            BigInteger n = key.PublicKey.Curve.N;
+            byte[] message = System.Text.Encoding.Default.GetBytes("HelloWorld");
+            MethodInfo dynMethod = typeof(ECDsa).GetMethod("CalculateE", BindingFlags.NonPublic | BindingFlags.Instance);
+            ((BigInteger)dynMethod.Invoke(sa, new object[] { n, message })).Should().Be(BigInteger.Parse("341881320659934023674980"));
```

Comment:
In my opinion, we should change this (and all of them) to `internal` instead of using reflection for testing

Reply:
I'm not quite clear how to use internal instead of using reflection. Could you please explain how to do it?
In addition, there are many places in the code that use reflection to call private and protected methods. Because the unit test does not need to consider performance, if the bad effects are not big, can we keep it?

Reply:
I think there is no need to test private methods. Because the complete test of public methods can theoretically cover all private methods. The parts that are not covered indicate that they are not being used and do not need to be tested.

Reply:
This Erik is a box of surprises. ahwuaha

Makes sense, Erik.

Reply:
I think we should make the complete test of private methods also.
Consider this case,  when one test path is too long, such as PublicMethodA --> PrivateMethodB---> PrivateMethodC---> ....... PrivateMethodN. 
If we want to test the PublicMethodA, there are so many conditions, which lead to a very complex unit test. But if we use bottom-up mode, it can simplify our unit test.

Reply:
I think there is a case where it is convenient to test with reflection. For example, a private method has several branches. There are more than one public method （maybe more than ten methods) for calling this private method, and each public method only call one branch in private method.  In this case if we use the public method as the entry to test It is quite cumbersome, and reflection can easily simulate input conditions.

Reply:
Since our goal is to fully cover public methods, there is no need to consider private methods.

Reply:
OK, we will remove theses reflection for private method test.

Reply:
Hi @erikzhang, maybe it is not needed to remove the current ones they already designed, no?
Just direct the focus of the future tests.

I agree with this comment also, Erik:
> "There are more than one public method （maybe more than ten methods) for calling this private method, and each public method only call one branch in private method."

Maybe, in a long-term, only test public is the desired thing. But maybe, for now, for simplicity, we could allow some private method to be tested until we reach a higher level of coverage.

What do you think, @igormcoelho?